extension/functions/src/presto.rs

```diff
@@ -62,39 +69,154 @@ impl ScalarFunctionDef for CurrentTimeFunction {
+    fn name(&self) -> &str {
+        "current_timestamp_p"
+    }
+
+    fn signature(&self) -> Signature {
+        Signature::exact(vec![DataType::Int64], Volatility::Immutable)
+    }
+
+    fn return_type(&self) -> ReturnTypeFunction {
+        Arc::new(move |_| Ok(Arc::new(DataType::Timestamp(TimeUnit::Nanosecond, None))))
+    }
+
+    fn execute(&self, args: &[ArrayRef]) -> Result<ArrayRef> {
+        let precision = args[0]
+            .as_any()
+            .downcast_ref::<Int64Array>()
+            .unwrap()
+            .value(0) as usize;
+
+        let now = SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .map_err(|err| DataFusionError::Execution(err.to_string()))?;
+        let nanos = now.as_nanos() as i64;
+
+        let adjusted_nanos = if precision > 9 {
+            nanos
+        } else {
+            let factor = 10_i64.pow(9 - precision as u32);
+            (nanos / factor) * factor
+        };
+
+        let array = TimestampNanosecondArray::from(vec![Some(adjusted_nanos)]);
+        Ok(Arc::new(array) as ArrayRef)
+    }
+}
 
 // Function package declaration
 pub struct FunctionPackage;
 
 impl ScalarFunctionPackage for FunctionPackage {
     fn functions(&self) -> Vec<Box<dyn ScalarFunctionDef>> {
-        vec![Box::new(CurrentTimeFunction)]
+        vec![
+            Box::new(CurrentTimeFunction),
+            Box::new(CurrentTimestampFunction),
+            Box::new(CurrentTimestampPFunction),
+        ]
     }
 }
 
 #[cfg(test)]
 mod test {
-    use chrono::Local;
+    use std::{
+        sync::Arc,
+        time::{SystemTime, UNIX_EPOCH},
+    };
+
+    use arrow::array::{
+        ArrayRef, Int64Array, TimestampMillisecondArray, TimestampNanosecondArray,
+    };
+    use chrono::Utc;
     use datafusion::error::Result;
     use datafusion::prelude::SessionContext;
+    use datafusion_expr::ScalarFunctionDef;
     use tokio;
 
-    use crate::utils::{execute, test_expression};
+    use crate::{
+        presto::{CurrentTimestampFunction, CurrentTimestampPFunction},
+        utils::{execute, test_expression},
+    };
 
     use super::FunctionPackage;
 
     #[tokio::test]
     async fn test_current_time() -> Result<()> {
-        let current = Local::now();
+        let current = Utc::now();
         let formatted = current.format("%H:%M:%S").to_string();
         test_expression!("current_time()", formatted);
         Ok(())
     }
+
+    #[tokio::test]
+    async fn test_current_timestamp() -> Result<()> {
+        let current_timestamp_function = CurrentTimestampFunction {};
+        let result = current_timestamp_function.execute(&[]).unwrap();
+        let result_array = result
+            .as_any()
+            .downcast_ref::<TimestampMillisecondArray>()
+            .unwrap()
+            .value(0);
+        let now = SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .unwrap()
+            .as_millis() as i64;
```

Comment:
Is this duplicate code with the test_current_timestamp_p()
If so, can you abstract a util function?

Reply:
Ok